### PR TITLE
DRILL-4606: Add DrillClient.Builder class

### PR DESCRIPTION
**Pull request by sudheeshkatkam**

### CODE REVIEW

**Comment by sudheeshkatkam** [comment]
@hnfgns can you please review?

**Comment by sudheeshkatkam** [comment]
Rebased on latest master.

@parthchandra and @adeneche please review?

File: exec/java-exec/src/main/java/org/apache/drill/exec/client/DrillClient.java [hunk 1]
```diff
@@ -73,81 +78,173 @@
...
+          @Override
+          protected void afterExecute(final Runnable r, final Throwable t) {
+            if (t != null) {
+              logger.error(String.format("%s.run() leaked an exception.", r.getClass().getName()), t);
+            }
+            super.afterExecute(r, t);
+          }
+        };
+
+    this.supportComplexTypes = builder.supportComplexTypes;
```
Comment (adeneche) [inline exec/java-exec/src/main/java/org/apache/drill/exec/client/DrillClient.java]: supportComplexTypes used to be initialized from the config object, and now it will be `true` by default unless we explicitly set it in the builder.

File: exec/java-exec/src/main/java/org/apache/drill/exec/client/DrillClient.java [hunk 2]
```diff
@@ -74,81 +78,173 @@
...
```
Comment (adeneche) [inline exec/java-exec/src/main/java/org/apache/drill/exec/client/DrillClient.java]: supportComplexTypes used to be initialized from the config object, and now it will be true by default unless we explicitly set it in the builder.

File: exec/java-exec/src/main/java/org/apache/drill/exec/client/DrillClient.java [hunk 3]
```diff
@@ -73,81 +78,173 @@
...
+
+    this.isDirectConnection = builder.isDirectConnection;
+    this.ownsZkConnection = builder.clusterCoordinator == null && !isDirectConnection;
+    this.clusterCoordinator = builder.clusterCoordinator; // could be null
+
+    this.ownsEventLoopGroup = builder.eventLoopGroup == null;
+    this.eventLoopGroup = !ownsEventLoopGroup ? builder.eventLoopGroup :
+        TransportCheck.createEventLoopGroup(config.getInt(ExecConstants.CLIENT_RPC_THREADS), "Client-");
+
+    this.ownsExecutor = builder.executor == null;
```
Comment (adeneche) [inline exec/java-exec/src/main/java/org/apache/drill/exec/client/DrillClient.java]: what is the motivation behind moving initiating `eventLoopGroup` and `executor` from `connect(String connect, Properties props)` to here, assuming the `connect` method was never called more than once ?

File: exec/java-exec/src/main/java/org/apache/drill/exec/server/BootStrapContext.java [hunk 4]
```diff
@@ -53,6 +54,8 @@ public BootStrapContext(DrillConfig config, ScanResult classpathScan) {
     this.loop2 = TransportCheck.createEventLoopGroup(config.getInt(ExecConstants.BIT_SERVER_RPC_THREADS), "BitClient-");
     // Note that metrics are stored in a static instance
     this.metrics = DrillMetrics.getRegistry();
+    this.userLoopGroup = TransportCheck.createEventLoopGroup(config.getInt(ExecConstants.USER_SERVER_RPC_THREADS),
```
Comment (adeneche) [inline exec/java-exec/src/main/java/org/apache/drill/exec/server/BootStrapContext.java]: shouldn't this change be part of a different pull request ? doesn't seem related to adding DrillClient.Builder

File: exec/java-exec/src/main/java/org/apache/drill/exec/client/DrillClient.java (hunk 2 above)
Comment (parthchandra) [inline exec/java-exec/src/main/java/org/apache/drill/exec/client/DrillClient.java]: This is probably OK. This is the Java client and the client does, in fact, support complex types.

File: exec/java-exec/src/main/java/org/apache/drill/exec/client/DrillClient.java (hunk 3 above)
Comment (parthchandra) [inline exec/java-exec/src/main/java/org/apache/drill/exec/client/DrillClient.java]: Not clear to me either. Why start the executor pool unless there is going to be a connection?

File: exec/java-exec/src/main/java/org/apache/drill/exec/client/DrillClient.java [hunk 5]
```diff
@@ -74,81 +78,173 @@
...
+    this(newBuilder()
+        .setConfig(config)
+        .setClusterCoordinator(coordinator)
+        .setAllocator(allocator));
+  }
+
+  /**
+   * @deprecated Create a DrillClient using {@link DrillClient.Builder}.
+   */
+  @Deprecated
```
Comment (parthchandra) [inline exec/java-exec/src/main/java/org/apache/drill/exec/client/DrillClient.java]: Usually, this is a good thing, deprecating a method you intend to remove. But in this case the number of ctors is getting too messy. Should we just remove them altogether?

Comment (parthchandra) [inline exec/java-exec/src/main/java/org/apache/drill/exec/client/DrillClient.java]: Hmm, I take that back, on reflection. Some apps might be using this class. Probably good idea to log a JIRA to remove these. in future.

File: exec/java-exec/src/main/java/org/apache/drill/exec/client/DrillClient.java (hunk 3 above)
Comment (sudheeshkatkam) [inline exec/java-exec/src/main/java/org/apache/drill/exec/client/DrillClient.java]: These constructors are light weight, no real work is done.

File: exec/java-exec/src/main/java/org/apache/drill/exec/server/BootStrapContext.java (hunk 4 above)
Comment (sudheeshkatkam) [inline exec/java-exec/src/main/java/org/apache/drill/exec/server/BootStrapContext.java]: Filed DRILL-4841.

**Comment by sudheeshkatkam** [comment]
@parthchandra please review.

**Comment by parthchandra** [comment]
+1

**Comment by vvysotskyi** [comment]
Looks like these changes were added to https://github.com/apache/drill/pull/565, so closing this PR.
